src/registry.c

```diff
@@ -0,0 +1,914 @@
+
+    if (!(flags & RXKB_CONTEXT_NO_DEFAULT_INCLUDES) &&
+        !rxkb_context_include_path_append_default(ctx)) {
+        rxkb_context_unref(ctx);
+        return NULL;
+    }
+
+    return ctx;
+}
+
+XKB_EXPORT bool
+rxkb_context_include_path_append(struct rxkb_context *ctx, const char *path)
+{
+    struct stat stat_buf;
+    int err;
+    char *tmp;
+
+    tmp = strdup(path);
+    if (!tmp)
+        goto err;
+
+    err = stat(path, &stat_buf);
+    if (err != 0)
+        goto err;
+    if (!S_ISDIR(stat_buf.st_mode))
+        goto err;
+
+#if defined(HAVE_EACCESS)
+    if (eaccess(path, R_OK | X_OK) != 0)
+        goto err;
+#elif defined(HAVE_EUIDACCESS)
+    if (euidaccess(path, R_OK | X_OK) != 0)
+        goto err;
+#endif
+
+    darray_append(ctx->includes, tmp);
+    return true;
+
+err:
+    return false;
+}
+
+XKB_EXPORT bool
+rxkb_context_include_path_append_default(struct rxkb_context *ctx)
+{
+    const char *home, *xdg, *root;
+    char *user_path;
+    int err;
+    bool ret = false;
+
+    home = secure_getenv("HOME");
+
+    xdg = secure_getenv("XDG_CONFIG_HOME");
+    if (xdg != NULL) {
+        err = asprintf(&user_path, "%s/xkb", xdg);
+        if (err >= 0) {
+            ret |= rxkb_context_include_path_append(ctx, user_path);
+            free(user_path);
+        }
+    } else if (home != NULL) {
+        /* XDG_CONFIG_HOME fallback is $HOME/.config/ */
+        err = asprintf(&user_path, "%s/.config/xkb", home);
+        if (err >= 0) {
+            ret |= rxkb_context_include_path_append(ctx, user_path);
+            free(user_path);
+        }
+    }
+
+    if (home != NULL) {
+        err = asprintf(&user_path, "%s/.xkb", home);
+        if (err >= 0) {
+            ret |= rxkb_context_include_path_append(ctx, user_path);
+            free(user_path);
+        }
+    }
+
+    root = secure_getenv("XKB_CONFIG_ROOT");
+    if (root != NULL)
+        ret |= rxkb_context_include_path_append(ctx, root);
+    else
+        ret |= rxkb_context_include_path_append(ctx, DFLT_XKB_CONFIG_ROOT);
+
+    return ret;
+}
+
+XKB_EXPORT bool
+rxkb_context_parse_default_ruleset(struct rxkb_context *ctx)
+{
+    return rxkb_context_parse(ctx, DEFAULT_XKB_RULES);
+}
+
+XKB_EXPORT bool
+rxkb_context_parse(struct rxkb_context *ctx, const char *ruleset)
+{
+    char **path;
+    bool success = false;
+
+    darray_foreach(path, ctx->includes) {
+        char rules[PATH_MAX];
+
+        snprintf(rules, sizeof(rules), "%s/rules/%s.xml", *path, ruleset);
+        success = parse(ctx, rules, RXKB_POPULARITY_STANDARD);
+
+        if (success) {
+            if (ctx->load_extra_rules_files) {
+                /* Where we find a directory with our wanted ruleset xml files
+                 * and that ruleset parses successfully, parse the .extras.xml
+                 * file as well (doesn't matter if it doesn't succeed).
+                 *
+                 * FIXME: should only do this for the system directories, I
```

Comment:
long-term we should get rid of the `extras.xml` file and use the existing but unused `popularity` tag instead (see the [xkeyboard-config merge request](https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/merge_requests/76)). We might never be able to get rid of this behaviour for system directories, but we can force the right thing for the custom includes.

and, tbh, I'm also having a hard time thinking of a use-case where you want to define keyboard layouts outside the system directories but you *don't* want them to show up in the default list.

Reply:
`gnome-desktop` uses dconf key `"org.gnome.desktop.input-sources` `show-all-sources` (toggle in `gnome-tweaks`) to decide whether to load it or not. I don't know how often it's used but generally if GNOME decides to expose something, someone's has asked for it :) I think say that's prominent enough and a large enough user base.

Either way, since we're the intermediate here it's much easier for us to just pass that detail on and let the caller decide whether to ignore it or not. And as I said, it won't matter for virtually all use-cases anyway.